File: app/lib/geocell.dart
/// PlotProof geocell utilities.
///
/// Snaps GPS coordinates to geohash cells and converts them to the
/// bytes32 cell ids the PlotProof contract uses as mapping keys.
///
/// Precision guide (cell size near the equator — Nigeria is close):
///   7 -> ~153m x 153m   (neighbourhood)
///   8 -> ~38m  x 19m    (plot scale)      <- default for PlotProof
///   9 -> ~4.8m x 4.8m   (spot scale)
///
/// A plot can straddle a cell boundary, so conflict checks must always
/// query the center cell AND its 8 neighbours (use [cellBlock]).
///
/// Pure Dart, no dependencies. The bytes32 encoding is the ASCII bytes
/// of the geohash string right-padded with zeros to 32 bytes — trivial
/// to reproduce in any language and cheap to compare on-chain.
library geocell;

import 'dart:typed_data';

const String _base32 = '0123456789bcdefghjkmnpqrstuvwxyz';
const int kPlotPrecision = 8;

/// Encode [lat], [lng] to a geohash string of [precision] characters.
String encodeGeohash(double lat, double lng, {int precision = kPlotPrecision}) {
  assert(lat >= -90 && lat <= 90, 'lat out of range');
  assert(lng >= -180 && lng <= 180, 'lng out of range');
  assert(precision > 0 && precision <= 22, 'precision out of range');

  double latMin = -90, latMax = 90;
  double lngMin = -180, lngMax = 180;

  final buf = StringBuffer();
  int bit = 0;
  int ch = 0;
  bool evenBit = true; // true = longitude bit, false = latitude bit

  while (buf.length < precision) {
    if (evenBit) {
      final mid = (lngMin + lngMax) / 2;
      if (lng >= mid) {
        ch = (ch << 1) | 1;
        lngMin = mid;
      } else {
        ch = ch << 1;
        lngMax = mid;
      }
    } else {
      final mid = (latMin + latMax) / 2;
      if (lat >= mid) {
        ch = (ch << 1) | 1;
        latMin = mid;
      } else {
        ch = ch << 1;
        latMax = mid;
      }
    }
    evenBit = !evenBit;

    if (++bit == 5) {
      buf.write(_base32[ch]);
      bit = 0;
      ch = 0;
    }
  }
  return buf.toString();
}

/// Decode a geohash to the (lat, lng) of its cell center.
({double lat, double lng}) decodeGeohash(String geohash) {
  final b = _bounds(geohash);
  return (
    lat: (b.latMin + b.latMax) / 2,
    lng: (b.lngMin + b.lngMax) / 2,
  );
}

/// Bounding box of a geohash cell. Useful for drawing the cell on a map.
({double latMin, double latMax, double lngMin, double lngMax}) boundsOf(
        String geohash) =>
    _bounds(geohash);

({double latMin, double latMax, double lngMin, double lngMax}) _bounds(
    String geohash) {
  double latMin = -90, latMax = 90;
  double lngMin = -180, lngMax = 180;
  bool evenBit = true;

  for (final c in geohash.toLowerCase().split('')) {
    final idx = _base32.indexOf(c);
    if (idx < 0) {
      throw ArgumentError('Invalid geohash character: $c');
    }
    for (int n = 4; n >= 0; n--) {
      final bit = (idx >> n) & 1;
      if (evenBit) {
        final mid = (lngMin + lngMax) / 2;
        if (bit == 1) {
          lngMin = mid;
        } else {
          lngMax = mid;
        }
      } else {
        final mid = (latMin + latMax) / 2;
        if (bit == 1) {
          latMin = mid;
        } else {
          latMax = mid;
        }
      }
      evenBit = !evenBit;
    }
  }
  return (latMin: latMin, latMax: latMax, lngMin: lngMin, lngMax: lngMax);
}

/// Adjacent cell in a direction: 'n', 's', 'e', 'w'.
///
/// Implemented by decoding to the cell center, stepping one cell-width
/// in the given direction, and re-encoding. Robust and easy to audit.
String neighbor(String geohash, String direction) {
  final b = _bounds(geohash);
  final latStep = b.latMax - b.latMin;
  final lngStep = b.lngMax - b.lngMin;
  var lat = (b.latMin + b.latMax) / 2;
  var lng = (b.lngMin + b.lngMax) / 2;

  switch (direction) {
    case 'n':
      lat += latStep;
    case 's':
      lat -= latStep;
    case 'e':
      lng += lngStep;
    case 'w':
      lng -= lngStep;
    default:
      throw ArgumentError('direction must be n/s/e/w');
  }

  // Wrap longitude; clamp latitude at the poles.
  if (lng > 180) lng -= 360;
  if (lng < -180) lng += 360;
  if (lat > 90 || lat < -90) return geohash; // no neighbour past a pole

  return encodeGeohash(lat, lng, precision: geohash.length);
}

/// The 3x3 block of cells around (and including) the cell containing
/// [lat], [lng]: [center, N, NE, E, SE, S, SW, W, NW].
///
/// This is what you pass to the contract's `claimCounts` /
/// `getClaimsBatch` so plots straddling a boundary still surface.
List<String> cellBlock(double lat, double lng,
    {int precision = kPlotPrecision}) {
  final c = encodeGeohash(lat, lng, precision: precision);
  final n = neighbor(c, 'n');
  final s = neighbor(c, 's');
  return <String>[
    c,
    n,
    neighbor(n, 'e'),
    neighbor(c, 'e'),
    neighbor(s, 'e'),
    s,
    neighbor(s, 'w'),
    neighbor(c, 'w'),
    neighbor(n, 'w'),
  ];
}

// -------------------------------------------------------------------
// bytes32 encoding (must match how the app writes cells on-chain)
// -------------------------------------------------------------------

/// Encode a geohash string as the bytes32 cell id used on-chain:
/// ASCII bytes, right-padded with zeros to 32 bytes.
Uint8List cellToBytes32(String geohash) {
  final codes = geohash.codeUnits;
  if (codes.length > 32) {
    throw ArgumentError('geohash too long for bytes32');
  }
  final out = Uint8List(32);
  out.setRange(0, codes.length, codes);
  return out;
}

/// Decode a bytes32 cell id back to its geohash string.
String bytes32ToCell(Uint8List b) {
  final end = b.indexOf(0);
  return String.fromCharCodes(b.sublist(0, end < 0 ? b.length : end));
}

/// Hex string (0x-prefixed) of a cell's bytes32 id — handy for
/// explorers and debugging.
String cellToHex(String geohash) {
  final b = cellToBytes32(geohash);
  final sb = StringBuffer('0x');
  for (final byte in b) {
    sb.write(byte.toRadixString(16).padLeft(2, '0'));
  }
  return sb.toString();
}

File: web/lib/geocell.ts
// Port of the app's geocell.dart so the website computes the SAME cells the
// contract uses. Geohash (base32), bytes32 = ASCII of the geohash right-padded.

const BASE32 = "0123456789bcdefghjkmnpqrstuvwxyz";
export const PLOT_PRECISION = 8;

export function encodeGeohash(
  lat: number,
  lng: number,
  precision = PLOT_PRECISION,
): string {
  let latMin = -90,
    latMax = 90,
    lngMin = -180,
    lngMax = 180;
  let out = "";
  let bit = 0;
  let ch = 0;
  let evenBit = true; // longitude bit first

  while (out.length < precision) {
    if (evenBit) {
      const mid = (lngMin + lngMax) / 2;
      if (lng >= mid) {
        ch = (ch << 1) | 1;
        lngMin = mid;
      } else {
        ch = ch << 1;
        lngMax = mid;
      }
    } else {
      const mid = (latMin + latMax) / 2;
      if (lat >= mid) {
        ch = (ch << 1) | 1;
        latMin = mid;
      } else {
        ch = ch << 1;
        latMax = mid;
      }
    }
    evenBit = !evenBit;
    if (++bit === 5) {
      out += BASE32[ch];
      bit = 0;
      ch = 0;
    }
  }
  return out;
}

type Bounds = { latMin: number; latMax: number; lngMin: number; lngMax: number };

function bounds(geohash: string): Bounds {
  let latMin = -90,
    latMax = 90,
    lngMin = -180,
    lngMax = 180;
  let evenBit = true;
  for (const c of geohash.toLowerCase()) {
    const idx = BASE32.indexOf(c);
    if (idx < 0) throw new Error(`Invalid geohash char: ${c}`);
    for (let n = 4; n >= 0; n--) {
      const b = (idx >> n) & 1;
      if (evenBit) {
        const mid = (lngMin + lngMax) / 2;
        if (b === 1) lngMin = mid;
        else lngMax = mid;
      } else {
        const mid = (latMin + latMax) / 2;
        if (b === 1) latMin = mid;
        else latMax = mid;
      }
      evenBit = !evenBit;
    }
  }
  return { latMin, latMax, lngMin, lngMax };
}

export function decodeGeohash(geohash: string): { lat: number; lng: number } {
  const b = bounds(geohash);
  return { lat: (b.latMin + b.latMax) / 2, lng: (b.lngMin + b.lngMax) / 2 };
}

export function neighbor(geohash: string, dir: "n" | "s" | "e" | "w"): string {
  const b = bounds(geohash);
  const latStep = b.latMax - b.latMin;
  const lngStep = b.lngMax - b.lngMin;
  let lat = (b.latMin + b.latMax) / 2;
  let lng = (b.lngMin + b.lngMax) / 2;
  if (dir === "n") lat += latStep;
  else if (dir === "s") lat -= latStep;
  else if (dir === "e") lng += lngStep;
  else lng -= lngStep;
  if (lng > 180) lng -= 360;
  if (lng < -180) lng += 360;
  if (lat > 90 || lat < -90) return geohash;
  return encodeGeohash(lat, lng, geohash.length);
}

/** center + 8 neighbours (the block a conflict check must cover). */
export function cellBlock(
  lat: number,
  lng: number,
  precision = PLOT_PRECISION,
): string[] {
  const c = encodeGeohash(lat, lng, precision);
  const n = neighbor(c, "n");
  const s = neighbor(c, "s");
  return [
    c,
    n,
    neighbor(n, "e"),
    neighbor(c, "e"),
    neighbor(s, "e"),
    s,
    neighbor(s, "w"),
    neighbor(c, "w"),
    neighbor(n, "w"),
  ];
}

/** Neighbours of an existing cell string (for conflict grouping). */
export function neighboursOf(cell: string): string[] {
  const n = neighbor(cell, "n");
  const s = neighbor(cell, "s");
  return [
    cell,
    n,
    neighbor(n, "e"),
    neighbor(cell, "e"),
    neighbor(s, "e"),
    s,
    neighbor(s, "w"),
    neighbor(cell, "w"),
    neighbor(n, "w"),
  ];
}

/** bytes32 hex (0x..) of a geohash: ASCII bytes right-padded to 32. */
export function cellToHex(geohash: string): `0x${string}` {
  const bytes = new Uint8Array(32);
  for (let i = 0; i < geohash.length && i < 32; i++) {
    bytes[i] = geohash.charCodeAt(i);
  }
  let hex = "0x";
  for (const b of bytes) hex += b.toString(16).padStart(2, "0");
  return hex as `0x${string}`;
}

/** Decode a bytes32 hex cell id back to its geohash string. */
export function hexToCell(hex: string): string {
  const h = hex.startsWith("0x") ? hex.slice(2) : hex;
  let s = "";
  for (let i = 0; i < h.length; i += 2) {
    const code = parseInt(h.slice(i, i + 2), 16);
    if (code === 0) break;
    s += String.fromCharCode(code);
  }
  return s;
}
